### PR TITLE
Refactor handlers and unit tests

### PR DESCRIPTION
**Pull request by zzat**
This PR contains a bunch of refactors:
- Move business logic from handlers to a separate namespace
- Change handler tests to focus on the http layer checks and verify if the correct function is invoked from model
- Switch to clj-http from ring-mock
-  Adding an exit-chan to signal worker to stop the async loop

### CODE REVIEW

File: swift-ticketing/src/swift_ticketing/model/ticket.clj [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            [swift-ticketing.db.booking :as db-booking]
+            [swift-ticketing.worker :as worker]))
+
+(defn create-tickets [db-spec uid event-id ticket-req]
+  (let [seat-type (:seat_type ticket-req)
+        tickets-map (if (= seat-type db-ticket/NAMED)
+                      (:seats ticket-req)
+                      (map (fn [_] {:name ""}) (range (:quantity ticket-req))))
+        price (:price ticket-req)
+        ticket-type-id (java.util.UUID/randomUUID)
```
[the remark this answers is not in the document]
Comment (olttwa) [reply swift-ticketing/src/swift_ticketing/model/ticket.clj:12]: Keep this cheatsheet handy: https://clojure.org/api/cheatsheet
First search for things there, then google, then Java Interop and if you don't find any implementations, roll-up your own :)

File: swift-ticketing/src/swift_ticketing/model/ticket.clj [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+(defn reserve-ticket [db-spec message-queue uid event-id booking-req]
+  (let [booking-id (java.util.UUID/randomUUID)
+        req-ticket-ids (:ticket_ids booking-req)
+        ticket-ids (when-not (nil? req-ticket-ids)
+                     (map #(java.util.UUID/fromString %) req-ticket-ids))]
+    (db-booking/insert-booking db-spec uid booking-id db-booking/INPROCESS)
+    (worker/add-reserve-ticket-request-to-queue
+     message-queue
+     {:booking-id booking-id
+      :ticket-type-id (:ticket_type_id booking-req)
```
Comment (olttwa) [inline swift-ticketing/src/swift_ticketing/model/ticket.clj:29]: I see a lot of snake-case keywords. The clojure convention is kebab-case. Can you convert these to kebab case? And JSON convention is camel-case. You can accept requests in camecase and convert them to kebab-case using coercion. Check if this makes sense: https://github.com/metosin/reitit/blob/master/doc/coercion/coercion.md?

File: swift-ticketing/src/swift_ticketing/worker.clj [hunk 3]
```diff
@@ -121,21 +121,28 @@
...
-          (println request)
-          (cond
-            (= event-type reserve-event) (handle-reserve-event db-spec redis-opts request)
-            (= event-type book-event) (handle-book-event db-spec request)
-            (= event-type cancel-event) (handle-cancel-event db-spec request)
-            :else (println "Worker: Unknown event")))
-        (catch Exception e
-          (println (str "Exception in Worker: " worker-id " :" e)))))))
+(defn process-ticket-requests [worker-id message-queue db-spec redis-opts exit-ch]
+  (async/go-loop []
```
Comment (olttwa) [inline swift-ticketing/src/swift_ticketing/worker.clj:125]: this function is too big and indented. Break it down into smaller chunks of work.

File: swift-ticketing/test/swift_ticketing/client.clj [hunk 4]
```diff
@@ -1,22 +1,35 @@
 (ns swift-ticketing.client
   (:require
-   [ring.mock.request :as mock]
+   [clj-http.client :as http]
    [clojure.data.json :as json]
-   [swift-ticketing.app :refer [swift-ticketing-app]]
    [swift-ticketing.fixtures :as fixtures]
    [swift-ticketing.factory :as factory]))
 
+(defn- uid-cookie [uid]
```
Comment (olttwa) [inline swift-ticketing/test/swift_ticketing/client.clj:8]: I'm not sure what's going on here. Can you elaborate?
If you're trying to do bi-directional routing, consider libraries like reitit or bidi?

File: swift-ticketing/test/swift_ticketing/client.clj [hunk 5]
```diff
@@ -26,81 +39,85 @@
...
-                      app)]
+   (let [response (get-request (str "/booking/" booking-id "/status"))]
+     {:path-params booking-id
+      :status (:status response)
+      :response (response-to-json response)})))
+
+(defn get-tickets
+  ([ticket-type-id]
+   (let [response (get-request "/ticket" {"ticket_type_id" ticket-type-id})]
+     (println response)
```
Comment (olttwa) [inline swift-ticketing/test/swift_ticketing/client.clj:113]: There are a lot of print statements scaterred across the code (either commented or not). Please clean those up

File: swift-ticketing/test/swift_ticketing/factory.clj [hunk 6]
```diff
@@ -64,11 +76,32 @@
...
+  {:ticket_id (java.util.UUID/randomUUID)
+   :ticket_name (mk-ticket-name (rand-int 1000))
+   :ticket_type_id (java.util.UUID/randomUUID)
+   :ticket_price (+ 10 (rand-int 10000))
+   :reservation_expiration_time (.plus (Instant/now)
+                                       (Duration/ofSeconds (+ 10 (rand-int 200))))
+   :ticket_status db-ticket/AVAILABLE
+   :booking_id (java.util.UUID/randomUUID)})
+
+(defn mk-tickets []
```
Comment (olttwa) [inline swift-ticketing/test/swift_ticketing/factory.clj:101]: This seems like a redundant function. You can remove this function and call map directly from the caller.
Checkout this blog: https://stuartsierra.com/2015/08/10/clojure-donts-redundant-map

File: swift-ticketing/test/swift_ticketing/client.clj [hunk 7]
```diff
@@ -1,22 +1,35 @@
 (ns swift-ticketing.client
```
Comment (olttwa) [inline swift-ticketing/test/swift_ticketing/client.clj:1]: Is a client inside a test considered a good practice for testing in Clojure?
@zzat Can you point me to any blogs where you have found this?
Also, did you consider using ring.mock to make these requests?

File: swift-ticketing/test/swift_ticketing/handlers_test.clj [hunk 8]
```diff
@@ -111,41 +81,184 @@
...
+    (let [{:keys [db-spec]} fixtures/test-env
+          booking-id (str (java.util.UUID/randomUUID))
+          expected-booking-status :booking-status
+          get-booking-status-called-with-correct-args (atom false)]
+      (with-redefs [booking/get-booking-status
+                    (fn [dbs bid]
+                      (when (and
+                             (= db-spec dbs)
+                             (= booking-id bid))
+                        (reset! get-booking-status-called-with-correct-args true))
```
Comment (olttwa) [inline swift-ticketing/test/swift_ticketing/handlers_test.clj:231]: This seems like a convoluted thing to do. You're putting a condition if something was called the way you expected it to be, and then setting an atom, and then putting assertions on the atom.
Why not put assertion on the condition itself?

**Review by olttwa — CHANGES_REQUESTED** [review]
Overall things look good, and a lot of progress from previous stage. However, a few minor-major changes are required.

File: swift-ticketing/src/swift_ticketing/model/ticket.clj [hunk 9]
```diff
@@ -0,0 +1,38 @@
+(ns swift-ticketing.model.ticket
+  (:require [swift-ticketing.db.ticket :as db-ticket]
+            [swift-ticketing.db.booking :as db-booking]
+            [swift-ticketing.worker :as worker]))
+
+(defn create-tickets [db-spec uid event-id ticket-req]
```
[the remark this answers is not in the document]
Comment (zzat) [reply swift-ticketing/src/swift_ticketing/model/ticket.clj:6]: I'm using [wrap-json-body](https://github.com/ring-clojure/ring-json/blob/0.5.1/src/ring/middleware/json.clj#L50) middleware with `keywords?` true. It converts json keys to keywords. As the key in json is `"seat_type"`, it is converted to `:seat_type`

File: swift-ticketing/src/swift_ticketing/handlers.clj [hunk 10]
```diff
@@ -1,170 +1,115 @@
...
+   [swift-ticketing.model.booking :as booking]))
 
-(defn respond-json [status body]
+(defn- respond-json [status body]
   {:status status
    :headers {"Content-Type" "application/json"}
    :body body})
 
-(defn respond-400 [body]
+(defn- respond-400 [body]
```
[the remark this answers is not in the document]
Comment (zzat) [reply swift-ticketing/src/swift_ticketing/handlers.clj:16]: Yeah I checked that out. Issue with that was that ring was setting `Content-Type` of responses to `application/octet-stream` as no explicit header is provided by [these](https://github.com/ring-clojure/ring/blob/master/ring-core/src/ring/util/response.clj#L47) functions.

File: swift-ticketing/src/swift_ticketing/specs.clj [hunk 11]
```diff
@@ -35,7 +35,7 @@
 (s/def ::seats vector?)
 (s/def ::ticket_ids vector?)
 
-(s/def ::get-event-params
```
[the remark this answers is not in the document]
Comment (zzat) [reply swift-ticketing/src/swift_ticketing/specs.clj:38]: Clojure's `uuid?` returns true if the arg passed is a `java.util.UUID`. In my case I have a `string` in json that is a `uuid`. Hence I need to call a `fromString` on the string to check if I am able to convert it into a `java.util.UUID`

File: swift-ticketing/src/swift_ticketing/model/ticket.clj (hunk 2 above)
Comment (zzat) [reply swift-ticketing/src/swift_ticketing/model/ticket.clj:29]: Before starting swift ticketing, I was going through [slack](https://api.slack.com/types/conversation) and [reddit](https://www.reddit.com/dev/api/) apis. They were using snake_case in json so I sticked to that. I'll switch to camelCase if that's the convention we use.
Will checkout coercion and convert all json keywords to kebab-case

File: swift-ticketing/test/swift_ticketing/client.clj (hunk 4 above)
Comment (zzat) [reply swift-ticketing/test/swift_ticketing/client.clj:8]: I'm creating a cookie here with the value as the `uid` passed to the function. The path `/` denotes that it is accessible to the entire domain. It is merely used for authentication and not for any redirection.

File: swift-ticketing/src/swift_ticketing/app.clj [hunk 12]
```diff
@@ -5,8 +5,7 @@
    [ring.middleware.json :refer [wrap-json-response wrap-json-body]]
    [swift-ticketing.handlers :as handlers]
    [compojure.api.sweet :as compojure]
-   [compojure.api.sweet :refer :all]
-   [swift-ticketing.app :as app]))
+   [compojure.api.sweet :refer :all]))
```
[the remark this answers is not in the document]
Comment (zzat) [reply swift-ticketing/src/swift_ticketing/app.clj]: Will push this change with a few other CI changes in main branch directly

**Review by olttwa — APPROVED** [review]

**Pull request merged**